Extensions/render/image.lisp

```diff
@@ -116,43 +116,117 @@
+                               (type image-dimension stencil-width stencil-height))
+               (let ((stencil-x (+ stencil-dx i))
+                     (stencil-y (+ stencil-dy j)))
+                 (setf alpha (if (and (<= 0 stencil-y stencil-height)
+                                      (<= 0 stencil-x stencil-width))
+                                 (aref stencil-array stencil-y stencil-x)
+                                 0)))))
+           (update-ink (i j)
+             (setf ink (clime:design-ink design i j))
+             (when (and (eq old-ink ink) (= old-alpha alpha))
+               (return-from update-ink))
+             (setf old-alpha alpha
+                   old-ink   ink)
+             (cond ((zerop alpha)
+                    (setf mode nil))
+                   ((typep ink 'standard-flipping-ink)
+                    (setf source-rgba (let ((d1 (slot-value ink 'climi::design1))
+                                            (d2 (slot-value ink 'climi::design2)))
+                                        (logand #x00ffffff
+                                                (logxor (climi::%rgba-value d1)
+                                                        (climi::%rgba-value d2)))))
+                    (if (= alpha 255)
+                        (setf mode :flipping)
+                        (setf source-a alpha
+                              mode     :flipping/blend)))
+                   ((= alpha 255)
+                    (let ((ink-rgba (climi::%rgba-value ink)))
+                      (if (= 255 (ldb (byte 8 24) ink-rgba))
+                          (setf source-rgba ink-rgba
+                                mode        :copy)
+                          (let-rgba ((r g b a) ink-rgba)
+                            (setf source-r r
+                                  source-g g
+                                  source-b b
+                                  source-a a
+                                  mode     :blend)))))
+                   (t ; If we get here, ALPHA is [1, 254].
+                    (locally (declare (type (integer 1 254) alpha))
+                      (let-rgba ((r.fg g.fg b.fg a.fg) (climi::%rgba-value ink))
+                        (setf source-r r.fg
+                              source-g g.fg
+                              source-b b.fg
+                              source-a (octet-mult a.fg alpha)
+                              ;; SOURCE-A is [0, 254], so never :COPY.
+                              mode     (if (zerop source-a)
+                                           nil
+                                           :blend))))))))
+      (do-regions ((src-j j y y y2)
+                   (src-i i x x x2))
+        (when (or (null clip-region)
+                  (region-contains-position-p clip-region src-i src-j))
+          (when stencil-array
+            (update-alpha i j))
+          (update-ink i j)
+          (case mode ; do nothing if MODE is NIL
```

Comment:
[optional] maybe add empty otherwise clause instead of a comment?

Reply:
I would use an `otherwise` clause if I wanted to emphasize the resulting value. In the case, I want to emphasize that nothing is done, so not having a clause seems better to me.